crates/edr_evm/src/blockchain.rs

```diff
@@ -130,10 +130,16 @@ pub trait Blockchain {
     /// Retrieves the hardfork specification used for new blocks.
     fn spec_id(&self) -> SpecId;
 
-    /// Retrieves the state at a given block
+    /// Retrieves the state at a given block.
+    ///
+    /// The state overrides are applied after the block they are associated
+    /// with. The specified override of a nonce may be ignored to maintain
+    /// validity.
     async fn state_at_block_number(
         &self,
         block_number: u64,
+        // Block number -> state overrides
+        state_overrides: &BTreeMap<u64, StateOverride>,
```

Comment:
When constructing a state at a block number, it's possible that someone wants to override the state at a particular block. An example of this are Hardhat's irregular state changes. These need to be passed in to ensure that they're applied in the same order.

E.g.

Genesis block: no transactions, so no state committed
Irregular: genesis accounts are added
Block 1: for all transactions combined, commit state
Block 2: for all transactions combined, commit state
Irregular: set_code & set_balance are called
Block 3: for all transactions combined, commit state

To retrieve the state, these need to be applied in the same order, including the state overrides (irregular state changes).

Previously we weren't doing this, resulting in incorrect state construction. This resolves state root inequality between HH & EDR, which we were seeing in the past.